### ISSUE TITLE
Suggestions from JOSS review for the Readme

### ISSUE THREAD

**Issue by mabesa** [mikejohnson51/AHGestimation#9]
Hi @mikejohnson51 
Great work, the mass-conserving fit looks like a very cool feature! I just got started on the review and have a few suggestions that may make the package easier to use internationally. I don't work in hydraulics so I hope you can forgive me if some of my comments or questions are basic or not relevant to your target users.  

Readme:
The readme is fully reproducible but not easy to understand for non-specialists. I have not gone through the vignettes yet but I noticed much more detailed explanations seem to be given there. Maybe you could link the vignettes to the readme. 

- I'd appreciate a more general introduction to what the package does. What does it mean that the package does Single Relation fits and Full hydraulic system fits?
- To a non-expert like me it would be helpful to have a section comparing your package to other rating curve fitting software. What are your target users? When should a user choose your package over other packages or methods? (e.g. I found these two methods: https://cran.r-project.org/web/packages/bdrc/bdrc.pdf, https://txwri.github.io/r-manual/stage-discharge.html) 
-> that would lead to a "statement of need" 
- Please add a section with installation instructions (fyi,I installed the software using the command devtools::install_github("mikejohnson51/AHGestimation") and updated all dependencies) 

Section Base data:
- I'd like to recommend explaining the variables in the 3 figures and their units. Please also consider adding the units in the data descriptor data.R as well as which columns are necessary and which columns are optional. As a side note on data.R: People not used to working with USGS data may have difficulty understanding where the data comes from. For me it is not a must but the data may be better understandable if you explain what the NWIS is and what the numeric code refers to. 

Section Single Relationship fit: 
- Please consider writing out OLS and NLS and maybe add a sentence of what ahg_estimate does in slightly more detail. Does it always fit using the two methods? Why did you choose these two methods? How do you recommend a user should select the better fit (in slightly more detail)? I just saw that you built in a recommendation based on the minimum nrmse, when verbose is set to TRUE. Maybe you could describe this as an example.
- Does the option allowance=0.05 do anything in the Single Relationship fit mode?

Section Full hydraulic fit:
- Here you seem to use the nsga2 method as well (according to my output when I run the readme) but in the readme you only talk about OLS and NLS. Would it make sense to introduce NSGA-II and to explain when your software uses which methods?
- I get the same results as you but from these results I do not understand the sentence below the figure. All methods perform similarly in terms of error (I don't see any difference of around 10% in the table).   

Section Hydaulic estimation: 
What does this mean, can the package estimate the shape of your cross-section from the fit? I'd like to ask for more information here.

**Comment by mikejohnson51** [comment]
Hi @mabesa! Thanks you so much for taking the time to write out all of these great points. I will start working through them and update this comment as I go. 

I will also notify you when complete! 

Thanks again for your willingness to help with this process. 

---- 

- [x] I'd appreciate a more general introduction to what the package does. What does it mean that the package does Single Relation fits and Full hydraulic system fits?

> We added an introduction section explaining the AHG equations and the need for mass conserving fits when looking at full hydraulic systems. We tried to highlight the terms of "single relationship" and "full hydraulic fits" with bolded text.

- [x] To a non-expert like me it would be helpful to have a section comparing your package to other rating curve fitting software. What are your target users? When should a user choose your package over other packages or methods? (e.g. I found these two methods: https://cran.r-project.org/web/packages/bdrc/bdrc.pdf, https://txwri.github.io/r-manual/stage-discharge.html)

-> that would lead to a "statement of need"

> We have added text to the introduction adding both of these examples and trying to enforce the idea that AHGs move beyond rating curves alone (single relationship) and tend to focus on the full system (while still providing the ability to compute single relationship fits)

- [x] Please add a section with installation instructions (fyi, I installed the software using the command `devtools::install_github("mikejohnson51/AHGestimation")` and updated all dependencies)

>  We have added a installation section at the top of the README using remotes.

## Section Base data:

- [x] I'd like to recommend explaining the variables in the 3 figures and their units. 

> The axis titles of all plots have been expanded to include units.

- [x] Please also consider adding the units in the data descriptor data.R as well as which columns are necessary and which columns are optional. 

>  We added units to the column names and description in the data.R file/documentation.

- [x] As a side note on `: People not used to working with USGS data may have difficulty understanding where the data comes from. For me it is not a must but the data may be better understandable if you explain what the NWIS is and what the numeric code refers to.

> We have added a file to `data-raw/nwis` that builds this data, it is also noted in the readme as a link. Hopefully the new description of the data (in the readme) also provides more clarity.

## Section Single Relationship fit:

- [x] Please consider writing out OLS and NLS and maybe add a sentence of what `ahg_estimate` does in slightly more detail. Does it always fit using the two methods? Why did you choose these two methods? How do you recommend a user should select the better fit (in slightly more detail)? 

> We spelled out the OLS and NLS acroymns and note that all results are rank ordered by nrmse in the text. We also tried to better introduce the function in

- [x] I just saw that you built in a recommendation based on the minimum nrmse, when verbose is set to TRUE. Maybe you could describe this as an example.

> The nrmse rankings should be returned regardless of `verbose` and this column has been described in more detail in the README text. To help make this clearer, we have changed all output columns to be `V_nrmse` or `total_nrmse` opposed to the previous `V_error` or `total_nrmse`. **This resulted in a refresh of the pkgdown site.**

- [x] Does the option allowance=0.05 do anything in the Single Relationship fit mode?

> The allowance parameters does not impact the single (or double) relationship fits. We have removed the explicit call to `allowance` in that example for clarity and added a more detailed description in the "Fitting hydraulic systems" section. 

## Section Full hydraulic fit:

- [x] Here you seem to use the `nsga2` method as well (according to my output when I run the readme) but in the readme you only talk about OLS and NLS. Would it make sense to introduce NSGA-II and to explain when your software uses which methods?

> We have tried to give a better brief description of this in the Readme and also link to the appropriate vignette.

- [x] I get the same results as you but from these results I do not understand the sentence below the figure. All methods perform similarly in terms of error (I don't see any difference of around 10% in the table).

> Thank you! That was left over from an example at a different gage. It has been removed.

## Section Hydaulic estimation:

- [x] What does this mean, can the package estimate the shape of your cross-section from the fit? I'd like to ask for more information here.

> We tried to expand on this section and added a link to the corresponding vignette.

**Comment by mikejohnson51** [comment]
Hi @mabesa! With the last PR from @arashmodrad I believe we are up to date with these changes. Thanks again!

**Comment by mabesa** [comment]
Great work!! It is much clearer what the package does. The examples and the literature list in the Readme give enough material to get one started with the algorithms used by the package. 

I'm going through the readme and the links therein and through the vignettes and leave further comments here below. 

**Readme:** 
- [x] Installation instructions: remotes::install_packages("mikejohnson51/AHGestimation") does not work. The command is remotes::install_github("mikejohnson51/AHGestimation")

**Vignettes:**
[The AGH relationship](https://mikejohnson51.github.io/AHGestimation/articles/traditonal-ahg.html)
- [x] Second paragraph in https://mikejohnson51.github.io/AHGestimation/articles/traditonal-ahg.html#traditonal-fitting-technique shows [38], maybe a missing link to a reference?

[Improving AHG estimates](https://mikejohnson51.github.io/AHGestimation/articles/improved-ahg.html)
- [x] Execution of line 176 results in error 'formal argument "xlab" matched by multiple actual arguments' 

[Fine tuning NSGA-2](https://mikejohnson51.github.io/AHGestimation/articles/optimize-nsga2.html)
- [x] Section [Flexibility](https://mikejohnson51.github.io/AHGestimation/articles/optimize-nsga2.html#flexibility) lists arguments to ahg_estimate but only explains the first. The arguments are described in the Reference so you could add a link to the Reference here. 

[Data filtering](https://mikejohnson51.github.io/AHGestimation/articles/data-filtering.html)
Only one typo:
- [x] Typo in first text line: steam instead of stream

[Hydraulic data generation](https://mikejohnson51.github.io/AHGestimation/articles/hydraulics.html)
- [x] Line 73 errors with "Error in curl::curl_fetch_disk(url, x$path, handle = handle) : 
  Failed to open file /Users/pkgbuild/Library/Application Support/org.R-project.R/R/nhdplusTools/nhdplusVAA.fst." 
- [x] Line 177: In round(n, 2) n is not defined and the line errors.

**Comment by mikejohnson51** [comment]
Thanks @mabesa for the continued guidance!

We resolved #11 (failing tests due to new test data columns); we added contributing guidelines; and now ensure we follow them!

We used `goodpractice:gp()` to fix the following:

1. Change `=` assignment operators to `<-`
2. Remove lines longer then 80 characters
3. Avoid using `1:length()` or `1:nrow()`
4. Suppress warnings in vignettes

**Comment by mikejohnson51** [comment]
All the above line items have been addressed! 

Thanks again @mabesa. We would also like to add you to the package as a reviewer if you would allow us?

**Note**: in the last issue raised, `n` is computed on line 79 via `compute_n(...)`

**Comment by mabesa** [comment]
Great!
I appreciate the thought but no need to mention me specifically. I think the published JOSS paper will mention reviewers. 

The following two issues still remain (on my laptop):

- For some reason I still get the error when plotting the figure under section NSGA-2 
[Improving AHG estimates](https://mikejohnson51.github.io/AHGestimation/articles/improved-ahg.html#nsga-ii)
Execution of line 176 results in error 'formal argument "xlab" matched by multiple actual arguments'. The rest of the vignette code runs without problem. 

- I also still get a curl error below. Am I missing a package or credentials? 
[Hydraulic data generation](https://mikejohnson51.github.io/AHGestimation/articles/hydraulics.html)
Line 73 errors with "Error in curl::curl_fetch_disk(url, x$path, handle = handle) :
Failed to open file /Users/pkgbuild/Library/Application Support/org.R-project.R/R/nhdplusTools/nhdplusVAA.fst."
Therefore I'm missing the variable n for line 177. 

**Parts of my session info**: 
R version 4.3.2 (2023-10-31)
Platform: aarch64-apple-darwin20 (64-bit)
Running under: macOS Sonoma 14.0

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] nhdplusTools_1.0.0 dplyr_1.1.4        patchwork_1.1.2    ggplot2_3.4.2      AHGestimation_0.2 

loaded via a namespace (and not attached):
 [1] tidyselect_1.2.0     Exact_3.2            viridisLite_0.4.1    rootSolve_1.8.2.4    arrow_14.0.0.2      
 [6] fastmap_1.1.1        RANN_2.6.1           digest_0.6.31        fst_0.9.8            fstcore_0.9.18      
[11] lifecycle_1.0.4      sf_1.0-12            lmom_3.0             magrittr_2.0.3       compiler_4.3.2      
[16] rlang_1.1.2          tools_4.3.2          utf8_1.2.4           yaml_2.3.7           data.table_1.14.10  
[21] knitr_1.42           bit_4.0.5            classInt_0.4-9       curl_5.2.0           xml2_1.3.4          
[26] KernSmooth_2.23-22   expm_0.999-8         withr_2.5.2          purrr_1.0.1          grid_4.3.2          
[31] fansi_1.0.6          e1071_1.7-14         colorspace_2.1-0     scales_1.2.1         MASS_7.3-60         
[36] cli_3.6.2            mvtnorm_1.2-4        rmarkdown_2.21       generics_0.1.3       rstudioapi_0.15.0   
[41] httr_1.4.7           readxl_1.4.3         gld_2.6.6            DBI_1.1.3            pbapply_1.7-2       
[46] proxy_0.4-27         stringr_1.5.0        dataRetrieval_2.7.14 rvest_1.0.3          assertthat_0.2.1    
[51] parallel_4.3.2       cellranger_1.1.0     vctrs_0.6.5          boot_1.3-28.1        webshot_0.5.5       
[56] Matrix_1.6-1.1       jsonlite_1.8.8       bit64_4.0.5          systemfonts_1.0.4    tidyr_1.3.0         
[61] units_0.8-2          glue_1.6.2           mco_1.15.6           stringi_1.7.12       hydroloom_1.0.1     
[66] gtable_0.3.3         munsell_0.5.0        tibble_3.2.1         pillar_1.9.0         htmltools_0.5.5     
[71] R6_2.5.1             evaluate_0.20        kableExtra_1.3.4     lattice_0.21-9       class_7.3-22        
[76] DescTools_0.99.52    Rcpp_1.0.12          zip_2.3.0            svglite_2.1.3        xfun_0.39           
[81] pkgconfig_2.0.3

**Comment by program--** [comment]
Hi @mabesa, Mike asked me to take a look at some of the issues popping up.

For the plotting error in *Improving AHG Estimates*, this is due to a bug in `mco:::plot.nsga2`. The current fix is to install the package `scatterplot3d`, which will be used in place of the base plotting functions within `plot.nsga2`. In order to fully fix this, we'll have to file an issue with the `mco` package.

**Comment by mikejohnson51** [comment]
@program-- thanks! What a rabbit hole that was. The key was that in the `mco` plot code, there is a switch, 
```
if (requireNamespace("scatterplot3d", quietly = TRUE)) {
            scatterplot3d::scatterplot3d(v, color = ifelse(o, 
                "red", "blue"))
        }
        else {
            pairs(v, col = col, pch = pch, ...)
 }
```

I had scatterplot3d installed, while I suspect @mabesa did not and thus the issue. 

To resolve this, the vignette now uses `scatterplot3d` and the new package has been added to the `DESCRIPTION` file as a Suggested package.


For your other issue, you would need to install the `VAA.fst` file using `nhdplusTools::download_vaa()`. This is overkill to highlight the utility so we just hardcoded the value for slope instead to save unneeded downloads. 

Thanks to both of you!!
